File: src/server/scrap/scrap.js
import chromium from 'playwright-aws-lambda'
import { PAGES_NEWS } from '../consts/const.js'

const cacheLinks = new Map()

async function scrapeLinks(url) {
	const browser = await chromium.launchChromium({ headless: true })
	const page = await browser.newPage()
	await page.goto(url)
	try {
		const links = await page.$$eval('a', linkElements => linkElements.map(link => link.href))
		return links.filter(
			href =>
				href.includes('.html') || href.includes('0') || href.includes('article')
		)
	} catch (error) {
		console.error('Error during scraping:', error)
		throw error
	} finally {
		await browser.close()
	}
}

// Función para mezclar aleatoriamente un arreglo (algoritmo de Fisher-Yates)
function randomLinks(links) {
	// Genera dos índices aleatorios diferentes
	const index1 = Math.floor(Math.random() * links.length)
	let index2
	do {
		index2 = Math.floor(Math.random() * links.length)
	} while (index2 === index1) // Asegura que los índices sean diferentes

	// Retorna los enlaces correspondientes a los índices generados
	return [links[index1], links[index2]]
}

// ------------> FUNCTION TO GET LINKS OF THE PAGES |-> EXPORT
export async function getLinksPage() {
	const index = Math.floor(Math.random() * PAGES_NEWS.length)
	const page = PAGES_NEWS[index]
	console.log(cacheLinks)
	let links = cacheLinks.get(page)

	if (!links) {
		try {
			links = await scrapeLinks(page)
			cacheLinks.set(page, links)
		} catch (error) {
			console.error('Error:', error)
			throw error
		}
	}

	return randomLinks(links)
}

export async function catchCards() {
	console.log('Catching data')
	for (let i = 0; i < PAGES_NEWS.length; i++) {
		try {
			const links = await scrapeLinks(PAGES_NEWS[i])
			cacheLinks.set(PAGES_NEWS[i], links)
		} catch (error) {
			console.error('Error:', error)
			throw error
		}
	}
	console.log('Finishing catching data')
}

File: src/server/scrap/scrap.py
from bs4 import BeautifulSoup
import random
import requests
from urllib.parse import urljoin


from consts.consts import PAGES_NEWS

cachelinks = {}


def scrape_links(url):
    try:
        # Send a GET request to the URL
        response = requests.get(url)
        # Parse the HTML content
        soup = BeautifulSoup(response.text, "html.parser")
        links = [urljoin(url, a["href"]) for a in soup.find_all("a", href=True)]
        links = [
            href
            for href in links
            if (
                (".html" in href 
                 or "0" in href 
                 or "article" in href)
                and not (
                    "video" in href
                    or "podcast" in href
                    or "season" in href
                    or "2024_presidential_campaign" in href
                )
            )
        ]
        return links[:5]
    except Exception as error:
        print("Error during scraping:", error)
        raise


# Function to randomly select two different links from a list
def random_links(links):
    index1 = random.randint(0, len(links) - 1)
    index2 = index1
    while index2 == index1:
        index2 = random.randint(0, len(links) - 1)
    return [links[index1], links[index2]]


# Function to get links from a page and cache them
def get_links_page():
    index = random.randint(0, len(PAGES_NEWS) - 1)
    page = PAGES_NEWS[index]
    links = cachelinks.get(page)
    if links is None:
        links = scrape_links(page)  # replace with your scraping function
        cachelinks[page] = links
    return random_links(links)


# Function to scrape links from all pages and cache them
def catch_cards():
    print("Catching data")
    for i in range(len(PAGES_NEWS)):
        try:
            links = scrape_links(PAGES_NEWS[i])  # replace with your scraping function
            cachelinks[PAGES_NEWS[i]] = links
        except Exception as error:
            print("Error:", error)
            raise
    print("Finishing catching data")
